### PR TITLE
Add podSecurityContext to Schedule and schedulable actions

### LINKED ISSUE #263: Ability to set SecurityContext for the various job types
## Summary

**As** a K8up user
**I want** to specify what user is used for the wrestic pods
**So that** it's able to backup all files

## Context

Wrestic 2.0 runs as root on purpose, there are some cases where PVCs can't be read otherwise.

But there also K8s distributions that do not want you to run the pods as root, like OpenShift. So we need a way to specify the `runAs` for the jobs so we can cover all the use cases.

Bonus: it would be great if the operator can determine this by itself. E.g. if the operator is running on OpenShift it will set a user and if running on another distro it will run as root.

## Out of Scope

* TBD

## Further links

* https://github.com/vshn/wrestic/pull/38
* https://github.com/vshn/wrestic/pull/19

## Acceptance criteria

*Given* a k8s cluster
*When* job is triggered
*Then* it will run with the right user

## Implementation Ideas

* none yet

### LINKED ISSUE #480: runAsUser, runAsGroup, fsGroup
## Summary

**As** an ops engineer setting up k8up\
**I want** control over the UID and GID with which (w)restic is started\
**So that** I can ensure tight security.

## Context

Currently restic is started as a specified user. This is either 1001 (65532 after #474) or root (maybe not anymore after #474), because these UIDs are burnt into the Docker image in the `Dockerfile`.

This could lead to problems with the mounts. I.e. if a Restore is performed, it will create the files with the permissions of the user it runs. When the container, who's files were restored, is restarted, it will see files with the UID 1001 (or 65532 or 0 respectively), which may prevent it's process to modify the files (edit, delete, maybe even read).

## Out of Scope

* Changes to other features

## Further links

* https://kubernetes.io/blog/2020/12/14/kubernetes-release-1.20-fsgroupchangepolicy-fsgrouppolicy/
* https://kubernetes.io/docs/tasks/configure-pod-container/security-context/

## Acceptance criteria

* It is possible to set the UID/GID with which the restore Pod is run.
* The GID of the mount `/restore` matches the GID with which the restore Pod is run.

## Implementation Ideas

* A global flag (e.g. env var, CLI flag, both) to define the default UID and GID for restore Pods, plus
* a flag in the Restore CRD to define the UID and GID under which the restore Pod is created.

### PR DESCRIPTION
**Pull request by cimnine**
## Summary

This PR aims to implement #263. A security context is added to the `Schedule` CRD, as well as to the definitions of all the schedulable actions.

Closes #263 .

## Checklist

- [x] Keep pull requests small so they can be easily reviewed.
- [x] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`,
      as they show up in the changelog
- [ ] Update the documentation.
- [x] Update tests.
- [x] Link this PR to related issues.

### CODE REVIEW

File: api/v1alpha1/job_object.go [hunk 1]
```diff
@@ -17,7 +17,10 @@ type JobObject interface {
 	GetType() JobType
 	// GetJobName returns the name of the underlying batch/v1 job.
 	GetJobName() string
+	// GetResources returns the specified resource requirements
 	GetResources() corev1.ResourceRequirements
+	// GetSecurityContext returns the specified security context
+	GetSecurityContext() *corev1.PodSecurityContext
```
Comment (ccremer) [inline api/v1alpha1/job_object.go:23]:
```suggestion
	// GetSecurityContext returns the specified pod security context
	GetSecurityContext() *corev1.PodSecurityContext
```

File: api/v1alpha1/check_types.go [hunk 2]
```diff
@@ -90,6 +90,11 @@ func (c *Check) GetResources() corev1.ResourceRequirements {
 	return c.Spec.Resources
 }
 
+// GetSecurityContext returns the security context
+func (c *Check) GetSecurityContext() *corev1.PodSecurityContext {
```
Comment (ccremer) [inline api/v1alpha1/check_types.go:94]: ...and all the other types that I'm too lazy to mark ;)
```suggestion
// GetSecurityContext returns the pod security context
func (c *Check) GetSecurityContext() *corev1.PodSecurityContext {
```

**Comment by cimnine** [comment]
> Do you guys know if pod security context is enough to be able to do backups as root? Or should it rather be container security context?

```yaml
apiVersion: backup.appuio.ch/v1alpha1
kind: Backup
metadata:
  name: demo-backup
spec:
  backend:
    ...
  podSecurityContext:
    runAsNonRoot: false
    runAsGroup: 0
    runAsUser: 0
```

The command runs as root on my testcluster:

```
2021-09-16T08:12:29.095Z	INFO	k8up	Starting k8up…	{"version": "snapshot", "date": "unknown", "commit": "unknown", "go_os": "linux", "go_arch": "amd64", "go_version": "go1.17", "uid": 0, "gid": 0}
```

The command does not run as root if `podSecurityContext.runAsNonRoot: true` or if no `podSecurityContext` is defined:

```
2021-09-16T08:45:56.903Z	INFO	k8up	Starting k8up…	{"version": "snapshot", "date": "unknown", "commit": "unknown", "go_os": "linux", "go_arch": "amd64", "go_version": "go1.17", "uid": 65532, "gid": 0}
```

**Review by Kidswiss — APPROVED** [review]
I believe that should be enough to backup the files as root.

Maybe it's possible to simulate that in some automated test? So we'd be sure if it works or not.

**Comment by cimnine** [comment]
> Maybe it's possible to simulate that in some automated test? So we'd be sure if it works or not.

The security context is defined in the Backup and the Restore CRD used for the e2e test. Without it, the backup would still work, but the restore fail, as the directory permissions would be wrong and restic would not be able to write to the directory.

**Pull request merged**